crates/runtime/src/trap_registry.rs

```diff
@@ -52,50 +89,77 @@ fn trap_code_to_expected_string(trap_code: ir::TrapCode) -> String {
     }
 }
 
-/// RAII guard for deregistering traps
-pub struct TrapRegistrationGuard(usize);
-
 impl TrapRegistry {
-    /// Registers a new trap.
-    /// Returns a RAII guard that deregisters the trap when dropped.
-    pub fn register_trap(
-        &mut self,
-        address: usize,
-        source_loc: ir::SourceLoc,
-        trap_code: ir::TrapCode,
-    ) -> TrapRegistrationGuard {
-        let entry = TrapDescription {
-            source_loc,
-            trap_code,
-        };
-        let previous_trap = self.traps.insert(address, entry);
-        assert!(previous_trap.is_none());
-        TrapRegistrationGuard(address)
-    }
+    /// Registers a list of traps.
+    ///
+    /// Returns a RAII guard that deregisters all traps when dropped.
+    pub fn register_traps(
+        &self,
+        list: impl IntoIterator<Item = (usize, ir::SourceLoc, ir::TrapCode)>,
+    ) -> TrapRegistration {
+        let mut start = usize::max_value();
+        let mut end = 0;
+        let mut traps = HashMap::new();
+        for (addr, source_loc, trap_code) in list.into_iter() {
+            traps.insert(
+                addr,
+                TrapDescription {
+                    source_loc,
+                    trap_code,
+                },
+            );
+            if addr < start {
+                start = addr;
+            }
+            if addr > end {
+                end = addr;
+            }
+        }
+        if traps.len() == 0 {
```

Comment:
API nit: perhaps `register_traps` should return `Option<TrapRegistration>` rather than a special no-op registration when no traps were registered (i.e. `TrapRegistration.end` should therefore be `usize` instead of `Option<usize>`)?

Reply:
I was pretty close to doing that in a previous version of this, but I realized that you could have multiple instances linked together, some with trap registrations and some without. If, while you're calling wasm, any one of them traps then we'd want to be able to recover the trap information for the original as well. 

Or put another way, `TrapRegistration` is how we do lookup of "what trap happened at this pc", so if `None` is returned we wouldn't be able to perform that lookup for some instances, but we probably want to be able to do it for all instances.

Reply:
I see what you're saying.  That makes sense.